Distribution Service Documentation (#870)
corona-warn-app/cwa-server: docs/DISTRIBUTION.md

## Diagnosis Key Export Files

At the end of the process the diagnosis keys are exported into ZIP archives that contain two entries:

- export.bin - the binary containing the exposure keys
- export.sig - the signature to verify the export binary

> The `export.bin` file of the archive contains the Temporary Exposure Keys that are broadcasted by the devices of the

Reviewer: suggested rewording "...that are uploaded from the devices of the ...."

## Signing

The signing of the files basically means creating and signing an archive that contain the `export.bin` file
(the file name is configurable in `application.yaml` under `tek-export.file-name` property).
Both the application configuration as well as the Temporary Exposure Keys are signed so that the mobile devices
can verify if the files are originating from the right backend. Using the signature, attack vectors like a
man in the middle attack, request forgery etc. can be mitigated successfully. The implementation is done in
[`AppConfigurationSigningDecorator`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/appconfig/structure/archive/decorator/signing/AppConfigurationSigningDecorator.java).

The signing process takes the `export.bin` archive and creates a signature of the data structure which is then written into the `export.sig`.
Please see`getSignatureFile` method located in [`SigningDecoratorOnDisk`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/structure/archive/decorator/signing/SigningDecoratorOnDisk.java).

> There are different implementations for that, i.e. how to sign an app configuration which is the

Reviewer: Because the export.bin archive contains different information (i.e. keys, app configuration), there are also different  implementations for signing each type of data structure:
- AppConfigurationSigningDecorator [link] / explanation
- DiagnosisKeySigningDecorator [link] / explanation
- TestSigningDecorator -> not sure exactly when this is used, or if used in prod environments

Author: The TestSigningDecorator it's used only in tests.

`AppConfigurationSigningDecorator` - the specific code to determine how an application configuration file is supposed to be signed.
For the Temporary Exposure Key files, they are signed in the similar way using the
[`DiagnosisKeySigningDecorator`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/structure/archive/decorator/signing/DiagnosisKeySigningDecorator.java).

The algorithm used for signing the archives as well as other relevant information is defined in the [`application.yaml`](/services/distribution/src/main/resources/application.yaml).

## Bundling and Shifting

The [`DiagnosisKeyBundler`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/DiagnosisKeyBundler.java)

> is responsible for the core logic of the distribution service, e.g. how the diagnosis keys are grouped into which date and which hour folder

Reviewer: "e.g. how the diagnosis keys are grouped into date and hour folders on the CDN"

and necessary transformations.

In the preparation step, the distribution job queries all diagnosis keys that are available in the database and

> cache them into the `DiagnosisKeyBundler`. The purpose of this class is to distribute keys into different hours and dates and tell

Reviewer: "The purpose of this class is to arrange keys based on their earliest time (date and hour) of distribution, and provide this information to the components which create the archives."

> cache them into the `DiagnosisKeyBundler`. The purpose of this class is to distribute keys into different hours and dates and tell

Reviewer 2: caches

the rest of the application for which hours and dates there are keys available.

> There are two of the key bundler, loaded based on the active profile.

Reviewer: "There are two implementations of the key bundler..."

It is noteworthy to mention that the productive diagnosis key bundler applies a series of policies on the keys originating from the country specified by the parameter`origin-country`, which is defined in the [`application.yaml`](/services/distribution/src/main/resources/application.yaml).  Furthermore, with the parameter `apply-policies-for-all-countries` it is possible to apply the policies to all keys. The first policy,

> which is described in the GAEN framework (more details [here](https://developer.apple.com/documentation/exposurenotification/setting_up_a_key_server)),

Reviewer 2: Policies are as follows:

- As described in the GAEN framework (more details [here](https://developer.apple.com/documentation/exposurenotification/setting_up_a_key_server)), diagnosis keys must not be distributed before two hours after the end of the key's expiration window.
- The shifting policy. Which ensures that each distribution / export file published to the CDN, contains at least 140 Temporary Exposure Keys (configurable by the `shifting-policy-threshold` property). When there are less keys available to distribute in a specific distribution run, these keys are shifted to a succeeding export, until the threshold minimum is fulfilled.

## Diagnosis Key Export Files

> At the end of the process the diagnosis keys are exported into ZIP archives that contains two entries:

Reviewer: "contain"

- export.sig - the signature to verify the export binary

The `export.bin` file of the archive contains the Temporary Exposure Keys that are broadcast by the defices of the
people who are diagnosed with COVID-19.
The `export.sig` file is used by the server to sign each `export.bin` with a private key. The `export.sig` file
contains the raw signature as well as additional information needed for verification.

## Signing

The signing of the files is basically creating an archive and that archive contains the `export.bin` file
(the file name is configurable in `application.yaml` under `tek-export.file-name` property).
Both the application configuration as well as the Temporary Exposure Keys are signed so that the mobile devices
can trust that this is coming from the right backend. There is no man in the middle attacks happening or
request forgery or any other attacks, so the server signs all of this. The way it is done is implemented in the
[`AppConfigurationSigningDecorator`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/appconfig/structure/archive/decorator/signing/AppConfigurationSigningDecorator.java).

> The signing of the archives is described by looking for an `export.bin` file. Within that archive, it will sign it or

Reviewer: Maybe a shorter sentence:
"The signing process takes the 'export.bin' archive and creates a signature of the data structure which is then written into the 'export.sig'. Please see getSignatureFile for details...

create a signature of that thing and basically just write it into an `export.sig` which is happening in the
`getSignatureFile` method located in [`SigningDecoratorOnDisk`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/structure/archive/decorator/signing/SigningDecoratorOnDisk.java).
There are different implementations for that, i.e. how to sign an app configuration which is the
AppConfigurationSigningDecorator - the specific code to determine how an application configuration file is supposed to be signed.
For the Temporary Exposure Key files, they are signed in the similar way using the
[`DiagnosisKeySigningDecorator`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/structure/archive/decorator/signing/DiagnosisKeySigningDecorator.java).

> The algorithm used for signing the archives as well as other relevant information is present in the [`application.yaml`](/services/distribution/src/main/resources/application.yaml).

Reviewer: Refer to the application.yml. for details on the specific algorithms used and other relevant information.

## Bundling and Shifting

The [`DiagnosisKeyBundler`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/DiagnosisKeyBundler.java)
is responsible fo most of the distribution magic like which diagnosis keys go into which date and which hour

> and any other transformations there must be applied.

Reviewer 3: ```suggestion
and necessary transformations.
```

> In the preparation step of the distribution it queries all diagnosis keys that are available in the database and put into

Reviewer: In the preparation step, the distribution job queries all ...and cache them into the 'DiagnosisKeyBundler'

this `DiagnosisKeyBundler`. The purpose of this class is to distribute keys into different hours and dates and tell
the rest of the application for which hours and dates there are keys available.

> There are two of those diagnosis key bundlers, which can basically be triggered by a specific profile being set.

Reviewer: There are two main implementations of the key bundler, loaded based on the active profile.

What is noteworthy for these bundlers is that there's a series of policies being applied here. The first policy,

> which is something that Apple demands (more details [here](https://developer.apple.com/documentation/exposurenotification/setting_up_a_key_server)),

Reviewer: I would refer to the 'GAEN framework' rather than Apple. I think the specification applies for Android as well, right?

Author: I couldn't find that information in the Google documentation.

Author: I changed it accordingly 😄

> which is something that Apple demands (more details [here](https://developer.apple.com/documentation/exposurenotification/setting_up_a_key_server)),
> expiry policy that states that mwe must not distribute a diagnosis key or a Temporary Exposure Key until at least

Reviewer 3: ```suggestion
which is described in the Apple framework (more details [here](https://developer.apple.com/documentation/exposurenotification/setting_up_a_key_server)),
is the expiry policy that states that diagnosis keys must not be distributed before
```

> two hours after the end of the key's expiration window. In other words, Apple wants us to wait an additional two hours
> before we distribute the key to further preserve the privacy of the users. Then, there's the second policy, and that is

Reviewer 3: ```suggestion
two hours after the end of the key's expiration window. 
Then, there's the second policy, which is called the shifting policy.
```

> being applied, which we call the shifting policy. We must make sure that for every distribution or for every export
> file that we put into the public, there must be at least 140 Temporary Exposure Keys in such an export

Reviewer 3: ```suggestion
It ensures that each distribution / export file that is published to the CDN, contains at least 140 Temporary Exposure Keys
```

> (the number is configurable by changing the `shifting-policy-threshold` property). If we cannot assemble enough or
> we do not have enough Temporary Exposure Keys for an export then we wait with the distribution of those keys until we
> have gathered enough to fulfill this threshold.

Reviewer 3: ```suggestion
(the number is configurable by changing the `shifting-policy-threshold` property). When there are less keys available
to distribute in a specific distribution run, these keys are shifted to a succeeding export, until the threshold of minimum
number of keys is fulfilled.
```

implementations for signing each type of data structure:

- [`AppConfigurationSigningDecorator`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/appconfig/structure/archive/decorator/signing/AppConfigurationSigningDecorator.java) - the specific code to determine how an application configuration file is supposed to be signed.
- [`DiagnosisKeySigningDecorator`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/structure/archive/decorator/signing/DiagnosisKeySigningDecorator.java) - used to sign Temporary Exposure Key files

The algorithm used for signing the archives as well as other relevant information is defined in the [`application.yaml`](/services/distribution/src/main/resources/application.yaml).

## Bundling and Shifting

The [`DiagnosisKeyBundler`](/services/distribution/src/main/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/DiagnosisKeyBundler.java)
is responsible for the core logic of the distribution service, e.g. how the diagnosis keys are grouped into date and

> hour filders on the CDN.

Reviewer 4: A typo: folders

Author: Done 😄

is responsible for the core logic of the distribution service, e.g. how the diagnosis keys are grouped into date and
hour folders on the CDN.

In the preparation step, the distribution job queries all diagnosis keys that are available in the database and
caches them into the `DiagnosisKeyBundler`. The purpose of this class is to arrange keys based on their earliest time
(date and hour) of distribution, and provide this information to the components which create the archives.
There are two implementations of the key bundler, loaded based on the active profile.
It is noteworthy to mention that the productive diagnosis key bundler applies a series of policies on the keys originating
from the country specified by the parameter`origin-country`, which is defined in the [`application.yaml`](/services/distribution/src/main/resources/application.yaml).
Furthermore, with the parameter `apply-policies-for-all-countries` it is possible to apply the policies to all keys.

Policies are as follows:

- As described in the GAEN framework (more details [here](https://developer.apple.com/documentation/exposurenotification/setting_up_a_key_server)),

> diagnosis keys must not be distributed before two hours after the end of the key's expeiration window.

Reviewer: Minor additional information, just in case you want to add it : "...after the end of the key's expeiration window (calculated by adding the rolling period of the key to its start interval number)"

Author: Done. 😄
